miniapps/blocksolvers/div_free_solver.cpp

```diff
@@ -0,0 +1,592 @@
+   data_.Q_l2.SetSize(num_refine);
+   hdiv_fes_->GetEssentialTrueDofs(ess_attr, data_.coarsest_ess_hdivdofs);
+   data_.C.SetSize(num_refine+1);
+
+   hcurl_fes_.reset(new ParFiniteElementSpace(mesh, hcurl_fec_.get()));
+   coarse_hcurl_fes_.reset(new ParFiniteElementSpace(*hcurl_fes_));
+   data_.P_hcurl.SetSize(num_refine, OperatorPtr(Operator::Hypre_ParCSR));
+}
+
+SparseMatrix* AggToInteriorDof(const Array<int>& bdr_truedofs,
+                               const SparseMatrix& agg_elem,
+                               const SparseMatrix& elem_dof,
+                               const HypreParMatrix& dof_truedof,
+                               Array<int>& agg_starts)
+{
+   OperatorPtr agg_dof(Mult(agg_elem, elem_dof));
+   SparseMatrix& agg_dof_ref = *agg_dof.As<SparseMatrix>();
+   OperatorPtr agg_tdof(dof_truedof.LeftDiagMult(agg_dof_ref, agg_starts));
+   OperatorPtr agg_tdof_T(agg_tdof.As<HypreParMatrix>()->Transpose());
+   SparseMatrix tdof_agg, is_shared;
+   HYPRE_Int* trash;
+   agg_tdof_T.As<HypreParMatrix>()->GetDiag(tdof_agg);
+   agg_tdof_T.As<HypreParMatrix>()->GetOffd(is_shared, trash);
+
+   int * I = new int [tdof_agg.NumRows()+1]();
+   int * J = new int[tdof_agg.NumNonZeroElems()];
+
+   Array<int> is_bdr;
+   FiniteElementSpace::ListToMarker(bdr_truedofs, tdof_agg.NumRows(), is_bdr);
+
+   int counter = 0;
+   for (int i = 0; i < tdof_agg.NumRows(); ++i)
+   {
+      bool agg_bdr = is_bdr[i] || is_shared.RowSize(i) || tdof_agg.RowSize(i)>1;
+      if (agg_bdr) { I[i+1] = I[i]; continue; }
+      I[i+1] = I[i] + 1;
+      J[counter++] = tdof_agg.GetRowColumns(i)[0];
+   }
+
+   double * D = new double[I[tdof_agg.NumRows()]];
+   std::fill_n(D, I[tdof_agg.NumRows()], 1.0);
+
+   SparseMatrix intdof_agg(I, J, D, tdof_agg.NumRows(), tdof_agg.NumCols());
+   return Transpose(intdof_agg);
+}
+
+void DFSSpaces::MakeDofRelationTables(int level)
+{
+   Array<int> agg_starts(Array<int>(l2_0_fes_->GetDofOffsets(), 2));
+   auto& elem_agg = (const SparseMatrix&)*l2_0_fes_->GetUpdateOperator();
+   OperatorPtr agg_elem(Transpose(elem_agg));
+   SparseMatrix& agg_el = *agg_elem.As<SparseMatrix>();
+
+   el_l2dof_.push_back(ElemToDof(*l2_fes_));
+   data_.agg_l2dof[level].Reset(Mult(agg_el, el_l2dof_[level+1]));
+
+   Array<int> bdr_tdofs;
+   hdiv_fes_->GetEssentialTrueDofs(all_bdr_attr_, bdr_tdofs);
+   auto tmp = AggToInteriorDof(bdr_tdofs, agg_el, ElemToDof(*hdiv_fes_),
+                               *hdiv_fes_->Dof_TrueDof_Matrix(), agg_starts);
+   data_.agg_hdivdof[level].Reset(tmp);
+}
+
+void DFSSpaces::CollectDFSData()
+{
+   auto GetP = [this](OperatorPtr& P, unique_ptr<ParFiniteElementSpace>& cfes,
+                      ParFiniteElementSpace& fes, bool remove_zero)
+   {
+      fes.Update();
+      fes.GetTrueTransferOperator(*cfes, P);
+      if (remove_zero)
+      {
+         hypre_ParCSRMatrixDropSmallEntries(*P.As<HypreParMatrix>(), 1e-16, -1);
```

Comment:
Until quite recently (including on my tuxbox) this hypre function only took two arguments, leading to the compilation error
```
div_free_solver.cpp:147:79: error: too many arguments to function ‘HYPRE_Int hypre_ParCSRMatrixDropSmallEntries(hypre_ParCSRMatrix*, HYPRE_Real)’
```
Since in MFEM we want to keep supporting older versions of hypre, this might need `#if MFEM_HYPRE_VERSION < 21400` or something similar around it.